### ISSUE TITLE
Add SVM as last layer of CNN

### ISSUE THREAD

**Issue by KodiaqQ** [keras-team/keras#9884]
Hello, all!
I have trained CNN with MLP at the end as multiclassifier.
Now, i want to try make this CNN without MLP (only conv-pool layers) to get features of image and get this features to SVM. Totally i think it will be CNN (keras) + SVM (sklearn) model, that i want to train at my dataset. How can i do this, to put together CNN features and SVM classifier? (maybe after flatten layer?) Thank you, and sry for bad eng.

**Comment by nuric** [comment]
You can create a model with just the CNN layers and use model.load_weights(fname, by_name=True) which will load only the available layers; you need to name your layers though (Conv(name="..")) Then you can just call model.predict to get NumPy arrays back as your CNN features and pass them into sklearn. You can flatten using NumPy before passing to sklearn or use any other preprocessing. Hope that helps.

**Comment by rex-yue-wu** [comment]
It actually makes more sense to me that you train everything in `keras`, because when you use `hinge_loss` to train a network, the last layer actually does the SVM job.

**Comment by shisigang** [comment]
FYI.
https://github.com/keras-team/keras/issues/2588

**Comment by gledsonmelotti** [comment]
How are you? I am trying to implement an SVM within of the keras cost function using sklearn.svm and sklearn.model_selection. However, I always get errors. I believe the problem is to convert the y_true and y_pred tensor into a numpy array to be used in sklearn.svm. Then I need to convert the predicted results to tensor to be used in the cost function of keras (categorical_hinge).

Can anybody help me?
`model_input = Input(shape = (img_width, img_height, channel_axis))`
`x = Convolution2D_bn(model_input, 32, 3, 3, strides=(2, 2), padding='valid')`
`x = Convolution2D_bn(x, 32, 3, 3, padding='valid')`
`...`
`x = GlobalAveragePooling2D()(x)`
`x = Dense(4096, kernel_regularizer=l2(1e-4), name='Dense_1')(x)`
`x = Activation('relu', name='relu1')(x)`
`x = Dropout(DROPOUT)(x)`
`x = Dense(4096, kernel_regularizer=l2(1e-4), name='Dense_2')(x)`
`x = Activation('relu', name='relu2')(x)`
`model_output = Dropout(DROPOUT)(x)`
`model = Model(model_input, model_output)`
`model.summary()`

`import tensorflow as tf`
`from keras import backend as K`
`from sklearn.svm import SVC`
`from sklearn.model_selection import GridSearchCV`
`from sklearn.preprocessing import StandardScaler`
`from keras.losses import categorical_hinge`
#
`def custom_loss_value(y_true, y_pred):`
   ` X = K.eval(y_pred)`
    `print(X)`
    `Y = np.ravel(K.eval(y_true))`
    `Predict = []`
    `Prob = []`
    `scaler = StandardScaler()`
    `X = scaler.fit_transform(X)`
    `param_grid = {'C': [0.1, 1, 8, 10], 'gamma': [0.001, 0.01, 0.1, 1]}`
    `SVM = GridSearchCV(SVC(kernel='rbf',probability=True), cv=3, param_grid=param_grid, scoring='auc', verbose=1)`
    `SVM.fit(X, Y)`
    `Final_Model = SVM.best_estimator_`
    `Predict = Final_Model.predict(X)`
    `Prob = Final_Model.predict_proba(X)`
    `return categorical_hinge(tf.convert_to_tensor(Y, dtype=tf.float32), tf.convert_to_tensor(Predict, dtype=tf.float32))`
#
`sgd = SGD(lr=0.001, decay=1e-6, momentum=0.9, nesterov=True)`
`model.compile(loss=custom_loss_value, optimizer=sgd, metrics=['accuracy'])`

**Comment by JanThan** [comment]
were you able to come to a solution?

**Comment by gledsonmelotti** [comment]
@JanThan I did not.
